docs/extended-collection-types.md

```diff
@@ -0,0 +1,169 @@
+# Extended Collection Types
+
+| Lead Author(s) | Implemented | GitHub Links |
+|---|---|---|
+| ike709 | :x: No | TBD |
+
+## Overview
+
+Currently, the only collection type in BYOND is the `list()` type, which doubles as both a dynamically resizable list and also as a dictionary. This RFC proposes exposing several additional collection types directly from C# to DM.
+
+Note that typed collections (e.g. an array of text strings) will be implemented as a future RFC; this RFC primarily focuses on *which* collection types to add in the first place.
+
+## Motivation
+
+The implementation of BYOND lists is highly inefficient in OpenDream as we can't just back it with a C# list or a C# dictionary. Furthermore, other collection types would offer performant convenience features for developers that they do not need to implement themselves, such as performant hashsets or stacks.
+
+## Design
+
+The current `list()` type will remain unchanged for BYOND compatibility, with the other types being added *in addition to* these legacy lists.
+
+The following collection types will essentially be exposed directly from C# to DM through a minimal wrapper API:
+
+- **Array**: A fixed-length collection of elements.
+- **Dictionary**: A collection of key-value pairs. Equivalent to existing associative lists, and will serve as the internal implementation of BYOND's planned `alist()` feature.
+- **HashSet**: A collection that contains no duplicate elements.
+- **ODList**: A dynamically resizable collection that is purely a list with none of the assoc-list functionality that BYOND lists have.
+- **Queue**: A first-in first-out (FIFO) collection.
+- **Stack**: A last-in first-out (LIFO) collection.
+
+Generally speaking, the helper procs from `list()` will also operate on other collections where applicable (e.g. `list.Join()` and `array.Join()` would be equivalent). A few collections will implement additional helpers as noted below.
+
+**Note:** Initializing OpenDream's collection types with a fixed size, for example `var/array/foo[10]`, will initialize it with a `capacity` of 10 but *will not* insert any `null` elements; the actual contents of the collection will be empty. This is **breaking behavior** from BYOND lists such as `var/list/L[10]`, which inserts 10 `null` elements into the list. BYOND lists in OpenDream will preserve this behavior for compatibility. Otherwise, see the "Collection Capacity API" subsection below for more information.
```

Comment:
How is it possible for a "fixed-length" collection of elements (the array) to have a `capacity` of 10 but contain no elements? I believe it should be populated with 10 nulls.

Reply:
> How is it possible for a "fixed-length" collection of elements (the array) to have a `capacity` of 10 but contain no elements? I believe it should be populated with 10 nulls.

This has to do with memory allocation, and the size of the memory allocated for the collection vs its actual contents. Capacity of 10 means it has space set aside for 10 elements, but may have any number of elements between 0 and 10.

This is what any modern language does, and DM itself likely does it internally without telling the user about it (allowing the user to see it means they can give the VM an advisory on how much memory they expect to need, and it can allocate it all ahead of time.)

Reply:
https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.capacity?view=net-8.0

Reply:
yeah but my point is that an array can not be resized, so it doesn't make sense for arrays specifically to have a capacity (at least, a capacity differing from its size)

Reply:
Thinking about this more, I really don't see how it makes sense for other containers either. Do any other languages exist with this kind of syntax for specifying the initial capacity of a container?

Is the `[10]` expected to be part of the type? I don't think that initial capacity should have anything to do with the type of the variable. I'd argue that this syntax should be entirely dropped for these new data types and moved to the right-hand-side of the variable declaration inside a default expression (somehow, I can't think of the nicest approach), where it probably belongs. This would also allow the feature to be used in places that are not variable declarations.

tbh BYOND's current behaviour makes far more sense than using this for capacity, although it's still weird.

The syntax maybe makes sense for non-reassignable array members, assuming it sets the size rather than initial capacity, but I don't think we need those.

Reply:
hm. I'm no longer sure either, can you bring this up in the discord?